Go Server and Dockerfile

dereknorrbom opened the pull request:
Implement HTTP server with TLS, handlers, and middleware

- Set up HTTP server with TLS configuration
- Implement GET health check endpoint (/healthz)
- Implement GET and POST methods for replica count
- Implement GET method for deployment listing
- Add logging middleware for request tracking
- Create handlers for replica count (/replica-count) and deployment listing (/deployments)
- Add error handling and JSON response encoding
- Ensure all endpoints have explicit method checks
- Add comprehensive unit tests for all handlers and utility functions (TDD)

TODO:
- Implement actual Kubernetes integration
- Consider separating handlers into their own package if complexity increases
- Evaluate need for additional middleware as the application grows

dereknorrbom committed:
Go Server and Dockerfile

Implement HTTP server with TLS, handlers, and middleware

- Set up HTTP server with TLS configuration
- Implement GET health check endpoint (/healthz)
- Implement GET and POST methods for replica count
- Implement GET method for deployment listing
- Add logging middleware for request tracking
- Create handlers for replica count (/replica-count) and deployment listing (/deployments)
- Add error handling and JSON response encoding
- Ensure all endpoints have explicit method checks
- Add comprehensive unit tests for all handlers and utility functions (TDD)

TODO:
- Implement actual Kubernetes integration
- Consider separating handlers into their own package if complexity increases
- Evaluate need for additional middleware as the application grows

taraspos commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,227 @@
+// listDeployments handles the /deployments endpoint to list deployments
+func listDeployments(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
```
While explicitly checking for HTTP request type in handler works just fine, it adds more boilerplate to the implementation. Any ideas about better way to handle this?

dereknorrbom replied:
I could use a middleware to achieve a DRY implementation while still checking the request method.

taraspos replied:
You could just handle it on the `mux` level, something like:

```
mux.HandleFunc("GET /healthz", healthCheck)
```

See some examples here:
- https://eli.thegreenplace.net/2023/better-http-server-routing-in-go-122

dereknorrbom replied:
Thank you for pointing that out, I went with your recommendation.

taraspos commented on build/package/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# Generate certificates
+COPY scripts ./scripts
+RUN chmod +x ./scripts/generate_certs.sh
+RUN sh ./scripts/generate_certs.sh
```
As far as I see `generate_certs.sh` generates both server and client certificates.
While it's clear why we need server certs inside container, do we need to keep client ones?

dereknorrbom replied:
The client certificates are not needed inside the server container.  I will update it to remove them from the server once generated and copied locally.

taraspos commented on build/package/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+# syntax=docker/dockerfile:1
+
```
```
➜  k8s-deployment-scaler git:(development/go-server) docker images k8s-deployment-scaler
REPOSITORY              TAG       IMAGE ID       CREATED          SIZE
k8s-deployment-scaler   latest    959ac20a826b   21 minutes ago   306MB
```

This is incredibly large image for this kind of application. Would be good to trim it down a bit.

dereknorrbom replied:
I will update it to use `scratch` and include only the necessary binary and certificates in the final image.

```
k8s-deployment-scaler dereknorrbom$ docker images k8s-deployment-scaler
REPOSITORY              TAG       IMAGE ID       CREATED         SIZE
k8s-deployment-scaler   latest    007c9d25ea3b   4 seconds ago   7.72MB
```

taraspos commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,227 @@
+
+	port := os.Getenv("SERVER_PORT")
+	if port == "" {
+		port = "8080"
```
`8080` more commonly used for HTTP servers. 8443 would be more common for this use-case.

dereknorrbom replied:
Thank you.  A leftover from prior to me implementing mTLS.  I will get that updated.

dereknorrbom committed:
Pull request updates from comments

- Update existing tests to use setupHandlers for consistent behavior
- Implement Go’s method-specific routing
- Use port 8443
- Reduce the final Docker image size
- Only copy the server certs to the final Docker image
- Ensure all endpoints return proper status codes and JSON responses
- Additional DRY improvements

dereknorrbom commented:
Updates have been pushed.

evanfreed reviewed (commented):
Could we make sure the Server code is feature complete before continuing? This way we can review that entire section in one pass.

For example, the requirements specify `HTTP API to get the list of available Deployments in the Kubernetes cluster`.  Right now getting the list of deployments isn't actually querying the Kubernetes API for anything.

dereknorrbom committed:
Kubernetes Integration

- Implement Kubernetes client integration in main.go
- Add deployment informer and cache for efficient data retrieval
- Update handlers to use Kubernetes client and cache
- Expand unit tests to cover new Kubernetes-related functions
- Expand Makefile to include Kubernetes-related build commands
- Add RBAC configuration (rbac.yaml) for necessary Kubernetes permissions
- Create Kubernetes deployment configuration (k8s-deployment-scaler.yaml)
- Dockerfile updates for efficiency and best practice of not using root user
- Update README

TODO for final PR:
- Implement Helm deployment
- Automation for Helm deployment

evanfreed reviewed (approved):
I think Taras covered most of my initial thoughts.  The current state looks better.

logand22 commented on build/package/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+RUN CGO_ENABLED=0 GOOS=linux go build -o k8s-deployment-scaler ./cmd/k8s-deployment-scaler
+
+# Stage 2: Create the final image
+FROM scratch
```
You might be interested in https://github.com/GoogleContainerTools/distroless which handles things like the creation of nonroot users. Or for the Go specific use-cases https://ko.build/

dereknorrbom replied:
Updated to use `gcr.io/distroless/static-debian11:nonroot`

logand22 commented on rbac.yaml:
```diff
@@ -0,0 +1,22 @@
+  name: deployment-reader-binding
+subjects:
+- kind: ServiceAccount
+  name: default
```
I recommend not using the default `ServiceAccount`.

dereknorrbom replied:
When implementing Helm I updated it to use a custom ServiceAccount

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,354 @@
+	if os.Getenv("KUBERNETES_SERVICE_HOST") != "" {
+		config, err = rest.InClusterConfig()
+	} else {
+		config, err = clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
```
This function actually falls back to the InClusterConfig when neither masterUrl or kubeconfigPath is passed in.

dereknorrbom replied:
Thank you, I have simplified it for the next PR.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,354 @@
+
+	// Set up deployment informer to watch for changes in deployments
+	factory := informers.NewSharedInformerFactory(clientset, time.Minute*10)
+	deploymentInformer = factory.Apps().V1().Deployments().Informer()
```
Usage of the lister might provide a better UX for you instead of the informer.

dereknorrbom replied:
The code now uses the lister.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,354 @@
+
+// DeploymentInfo stores the relevant information about a Deployment
+type DeploymentInfo struct {
+	Replicas   int32
```
Are replicas not a part of the deployment already? If so, doesn't this make this struct unnecessary

dereknorrbom replied:
YAGNI. Thank you.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,354 @@
+
+	// Update the deployment
+	info.Deployment.Spec.Replicas = &reqBody.Replicas
+	_, err := clientset.AppsV1().Deployments(namespace).Update(context.TODO(), info.Deployment, metav1.UpdateOptions{})
```
Is there a better context you can use here?

dereknorrbom replied:
You're right. Context matters. I'll use the request's context with a timeout. It's cleaner and safer. Thanks for catching that.

logand22 commented on cmd/k8s-deployment-scaler/main.go:
```diff
@@ -0,0 +1,354 @@
+	// Use the informer's cache to list deployments
+	for _, obj := range deploymentInformer.GetIndexer().List() {
+		deployment := obj.(*appsv1.Deployment)
+		if namespace == "" || deployment.Namespace == namespace {
```
Use of lister might make this easier.

dereknorrbom replied:
The code now uses the lister.
